README
NAME
    Fennec - A testers toolbox, and best friend

DESCRIPTION
    Fennec ties together several testing related modules and enhances their
    functionality in ways you don't get loading them individually. Fennec
    makes testing easier, and more useful.

SYNOPSYS
    There are 2 ways to use Fennec. You can use Fennec directly, or you can
    use the shiny sugar-coated interface provided by the add-on module
    Fennec::Declare.

  VANILLA SYNTAX
    If Devel::Declare and its awesome power of syntax specification scares
    you, you can always write your Fennec tests in the stone age like
    this... just don't miss any semicolons.

    t/some_test.t: package TEST::SomeTest; use strict; use warnings;

        use Fennec(
            parallel  => 3,
            test_sort => 'random',
        );

        # This is optional, there is a default 'new' if you do not override it.
        sub new { ... }

        # Test blocks are called as methods on an instance of your test package.
        tests group_1 => sub {
            my $self = shift;
            ok( 1, "1 is true" );
        };

        test group_2 => (
            todo => 'This is not ready yet',
            code => sub {
                my $self = shift;
                ok( 0, "Not ready" );
            }
        );

        # It is important to always end a Fennec test with this function call.
        done_testing();

  DECLARE SYNTAX
    Note: In order to use this you MUST install Fennec::Declare which is a
    separate distribution on cpan. This module is separate because it uses
    the controversial Devel::Declare module.

    t/some_test.t: package TEST::SomeTest; use strict; use warnings;

        use Fennec::Declare(
            parallel  => 3,
            test_sort => 'random',
        );

        # This is optional, there is a default 'new' if you do not override it.
        sub new { ... }

        # Test blocks are called as methods on an instance of your test package.
        tests group_1 {
            # Note: $self is automatically shifted for you.
            ok( $self, "Got self automatically" );
        };

        test group_2 ( todo => 'This is not ready yet' ) {
            # Note: $self is automatically shifted for you.
            ok( 0, "Not ready" );
        }

        # It is important to always end a Fennec test with this function call.
        done_testing;

FEATURES
  PROVIDED DIRECTLY BY FENNEC
    Forking just works
        Forking in perl tests that use Test::Builder is perilous at best.
        Fennec initiates an Fennec::Collector class which sets up
        Test::Builder to funnel all test results to the main thread for
        rendering. A result of this is that forking just works.

    Concurrency, test blocks can run in parallel
        By default all "test" blocks are run in parallel with a cap of 3
        concurrent processes. The process cap can be set with the "parallel"
        import argument.

    No need to maintain a test count
        The test count traditionally was used to ensure your file finished
        running instead of exiting silently too early. With Test::Builder
        and friends this has largely been replaced with the "done_testing()"
        function typically called at the end of tests. Fennec shares this
        concept, but takes it further, you MUST call "done_testing()" at the
        end of your test files. This is safer because it can be used to
        ensure your test script ran completely.

    Can be decoupled from Test::Builder
        Fennec is built with the assumption that Test::Builder and tools
        built from it will be used. However custom Fennec::Collector and
        Fennec::Runner classes can replace this assumption with any testing
        framework you want to use.

    Can run specific test blocks, excluding others
        Have you ever had a huge test that took a long time to run? Have you
        ever needed to debug a failing test at the end of the file? How many
        times did you need to sit through tests that didn't matter?

        With Fennec you can specify the "FENNEC_TEST" environment variable
        with either a line number or test block name. Only tests defined on
        that line, or with that name will be run.

    Predictability: Rand is always seeded with the date
        Randomizing the order in which test blocks are run can help find
        subtle interaction bugs. At the same time if tests are always in
        random order you cannot reliably reproduce a failure.

        Fennec always seeds rand with the current date. This means that on
        any given date the test run order will always be the same. However
        different days test different orders. You can always specify the
        "FENNEC_SEED" environment variable to override the value used to
        seed rand.

    Diag output is coupled with test output
        When you run a Fennec test with a verbose harness (prove -v) the
        diagnostic output will be coupled with the TAP output. This is done
        by sending both output to STDOUT. In a non-verbose harness the
        diagnostics will be sent to STDERR per usual.

    Works with Moose
        All your test classes are instantiated objects. You can use Moose to
        define these test classes. But you do not have to, you are not
        forced to use OOP in your tests.

  PROVIDED BY MODULES LOADED BY FENNEC
    The 3 most common and useful Test::* modules
        Test::More, Test::Warn, Test::Exception

    RSPEC support
        Those familiar with Ruby may already know about the RSPEC testing
        process. In general you "describe" something that is to be tested,
        then you define setup and teardown methods ("before_all",
        "before_each", "after_all", "after_each") and then finally you test
        "it". See the "EXAMPLES" section or Test::Workflow for more details.

    Test re-ordering, tests can run in random, sorted, or defined order.
        When you load Fennec you can specify a test order. The default is
        random. You can also use the order in which they are defined, or
        sorted (alphabetically) order. If necessary you can pass in a
        sorting function that takes a list of all test-objects as arguments.

        *Provided by Test::Workflow*

    Reusable test modules
        You can write tests in modules using Test::Workflow and then import
        those tests into Fennec tests. This is useful if you have tests that
        you want run in several, or even all test files.

        *Provided by Test::Workflow*

    Incredibly powerful mocking with a simple API
        You can create classless object instances from a specification on
        the fly, define new packages, or override existing packages.

        *Provided by Mock::Quick*

DEFAULT IMPORTED MODULES
    Note: These can be overridden either on import, or by subclassing
    Fennec.

    Child - Forking for dummies
        Child is an OO interface to forking that removes all the
        boilderplate such as checking if the pid changed, and making sure
        you exit the child process.

    Mock::Quick - Mocking without the eye gouging
        Mock::Quick is a mocking library that makes mocking easy. In additon
        it uses a declarative style interface. Unlike most other mocking
        libraries on CPAN, it does not make people want to gouge their eyes
        out and curl up in the fetal position.

    Test::Workflow - RSPEC for perl.
        Test::Workflow is a testing library written specifically for Fennec.
        This library provides RSPEC workflow functions and structure. It can
        be useful on its own, but combined with Fennec it gets concurrency.

    Test::More
        Tried and True testing module that everyone uses.

    Test::Warn
        Test::Warn - Test code that issues warnings.

    Test::Exception
        Test::Exception - Test code that throws exceptions

IMPORT ARGUMENTS
    base => 'Some::Base'
        Load the specified module and make it the base class for your test
        class.

    class => 'What::To::Test'
        Used to specify the name of the package your test file is
        validating. When this parameter is specified 3 things are done for
        you: The class is automatically loaded, the $CLASS variable is
        imported and contains the module name, and the class() subroutine is
        defined and returns the name.

            use Fennec class => 'Foo::Bar';

            ok( $INC{'Foo/Bar.pm'}, "Loaded 'Foo::Bar'" );
            is( $CLASS, 'Foo::Bar', "We have \$CLASS" );
            is( class(), 'Foo::Bar', "We have class()" );

            tests method => sub {
                my $self = shift;
                is( $self->class(), 'Foo::Bar', "We have class() method" );
            };

            done_testing;

    parallel => $PROC_LIMIT
        How many test blocks can be run in parallel. Default is 3. Set to 1
        to fork for each test, but only run one at a time. Set to 0 to
        prevent forking.

        You can also set this using the $FENNEC_PARALLEL environment
        variable.

    debug => 1
        Enable tracking debugging information. At the end of the Fennec run
        it will present you with a CSV temp file. This file lists all blocks
        that are run, and mocks that are made in sequence from top to
        bottom. The actions are split into columns by PID. This is usedul
        when debugging potential race-conditions when using parallel
        testing.

        Example:

            26150,26151,26152,26153,26154
            0 26150 BLOCK 54->78 child: outer_wrap, , , , , 
             ,1 26151 BLOCK 47->52 test: class_store, , , , 
            0 26150 MOCK Foo => (outer), , , , , 
            0 26150 BLOCK 58->61 before_all: ba, , , , , 
             , ,2 26152 MOCK Foo => (outer), , , 
             , ,2 26152 BLOCK 63->66 before_each: be, , , 
             , ,2 26152 BLOCK 68->72 test: the_check, , , 
             , , ,3 26153 BLOCK 16->31 test: object, , 
             , , , ,4 26154 BLOCK 33->45 test: class,

        You can use this in a spreadsheet program, or use this command to
        look at it in a more friendly way.

            column -s, -t < '/path/to/tempfile' | less -#2 -S

    collector_class => 'Fennec::Collector::TB::TempFiles'
        Specify which collector to use. Defaults to a Test::Builder based
        collector that uses temp files to funnel tests from child procs to
        the parent.

        You generally won't need to specify this, unless you use a test
        infrastructure that is neither TAP nore Test::Builder based.

    runner_class => 'Fennec::Runner'
        Specify the runner class. You probably don't need this.

    runner_params => { ... }
        Lets you specify arguments used when Fennec::Runner is initialized.

    skip_without => [ 'Need::This', 'And::This' ]
        Tell Fennec to skip the test file if any of the specified modules
        are missing.

    test_sort => $SORT
        Options: 'random', 'sorted', 'ordered', or a code block.

        Code block accepts a list of Test::Workflow::Test objects.

    utils => [ 'Test::Foo', ... ]
        Load these modules instead of the default list.

        If you need to specify import arguments for any specific util class,
        you can use the class name as the key with an arrayref containing
        the arguments.

            use Fennec(
                utils          => [ 'Some::Module' ],
                'Some::Module' => [ arg => $val, ... ],
            );

    with_tests => [ 'Reusable::Tests', 'Common::Tests' ]
        Load these modules that have reusable tests. Reusable tests are
        tests that are common to multiple test files.

    seed => '...'
        Set the random seed to be used. Defaults to current date, can be
        overridden by the FENNEC_SEED environment variable.

    debug => $BOOL
        Can be used to turn on internal debugging for Fennec. This currently
        does very little.

ENVIRONMENT VARIABLES
    FENNEC_SEED
        Can be used to set a specific random seed

    FENNEC_TEST
        Can be used to tell Fennec to only run specific tests (can be given
        a line number or a block name).

    FENNEC_DEBUG
        When true internal debugging is turned on.

EXPORTED FUNCTIONS
  FROM FENNEC
    done_testing()
    done_testing(sub { ... })
        Should be called at the end of your test file to kick off the RSPEC
        tests. Always returns 1, so you can use it as the last statement of
        your module. You must only ever call this once per test file.

        Never put tests below the done_testing call. If you want tests to
        run AFTER the RSPEC workflow completes, you can pass done_testing a
        coderef with the tests.

            done_testing( sub {
                ok( 1, "This runs after the RSPEC workflow" );
            });

  FROM Test::Workflow
    See Test::Workflow or "EXAMPLES" for more details.

    with_tests 'Module::Name';
        Import tests from a module

    tests $name => sub { ... };
    tests $name => ( %params );
    it $name => sub { ... };
    it $name => ( %params );
        Define a test block

    describe $name => sub { ... };
        Describe a set of tests (group tests and setup/teardown functions)

    case $name => sub { ... };
        Used to run a set of tests against multiple conditions

    before_all $name => sub { ... };
        Setup, run once before any tests in the describe scope run.

    before_each $name => sub { ... };
        Setup, run once per test, just before it runs.

    around_each $name => sub { ... };
        Setup and/or teardown.

    after_each $name => sub { ... };
        Teardown, run once per test, after it finishes.

    after_all $name => sub { ... };
        Teardown, run once, after all tests in the describe scope complete.

  FROM Mock::Quick
    See Mock::Quick or "EXAMPLES" for more details.

    my $control = qclass $CLASS => ( %PARAMS, %OVERRIDES );
    my $control = qtakeover $CLASS => ( %PARAMS, %OVERRIDES );
    my $control = qimplement $CLASS => ( %PARAMS, %OVERRIDES );
    my $control = qcontrol $CLASS => ( %PARAMS, %OVERRIDES );
        Used to define, takeover, or override parts of other packages.

    my $obj = qobj( %PARAMS );
    my ( $obj, $control ) = qobjc( %PARAMS );
    my $obj = qstrict( %PARAMS );
    my ( $obj, $control ) = qstrictc( %PARAMS );
        Define an object specification, quickly.

    my $clear = qclear();
        Used to clear a field in a quick object.

    my $meth = qmeth { ... };
    my $meth = qmeth( sub { ... } );
        Used to define a method for a quick object.

  OTHER
    See Test::More, Test::Warn, and Test::Exception

EXAMPLES
    Examples can be the best form of documentation.

  SIMPLE
   VANILLA SYNTAX
    t/simple.t

        use strict;
        use warnings;

        use Fennec;

        use_ok 'Data::Dumper';

        tests dumper => sub {
            my $VAR1;
            is_deeply(
                eval Dumper({ a => 1 }),
                { a => 1 },
                "Serialize and De-Serialize"
            );
        };

        tests future => (
            todo => "Not ready yet",
            code => sub {
                ok( 0, "I still have to write these" );
            },
        );

        done_testing;

   DECLARE SYNTAX
    t/simple.t

        use strict;
        use warnings;

        use Fennec::Declare;

        use_ok 'Data::Dumper';

        tests dumper {
            my $VAR1;
            is_deeply(
                eval Dumper({ a => 1 }),
                { a => 1 },
                "Serialize and De-Serialize"
            );

            is(
                eval { no strict; Dumper( { a => 1 } ) },
                { a => 1 },
                "Serialize and De-Serialize"
            );
        }

        tests future( todo => "Not ready yet" ) {
            ok( 0, "I still have to write these" );
        }

        done_testing;

  RUN TESTS UNDER DIFFERENT CONDITIONS
    This example shows 4 conditions ($letter as 'a', 'b', 'c', and 'd'). It
    also has 2 test blocks, one that verifies $letter is a letter, the other
    verifies it is lowercase. Each test block will be run once for each
    condition, 2*4=8, so in total 8 tests will be run.

   VANILLA
    sample.t:

        use strict;
        use warnings;

        use Fennec;

        my $letter;
        case a => sub { $letter = 'a' };
        case b => sub { $letter = 'b' };
        case c => sub { $letter = 'c' };
        case d => sub { $letter = 'd' };

        tests is_letter => sub {
            like( $letter, qr/^[a-z]$/i, "Got a letter" );
        };

        tests is_lowercase => sub {
            is( $letter, lc( $letter ), "Letter is lowercase" );
        };

        done_testing;

   OBJECT ORIENTED
    sample.t

        use strict;
        use warnings;

        use Fennec;

        sub letter {
            my $self = shift;
            ( $self->{letter} ) = @_ if @_;
            return $self->{letter};
        }

        describe letters => sub {
            case a => sub { shift->letter('a') };
            case b => sub { shift->letter('b') };
            case c => sub { shift->letter('c') };
            case d => sub { shift->letter('d') };

            tests is_letter => sub {
                my $self = shift;
                like( $self->letter, qr/^[a-z]$/i, "Got a letter" );
            };

            tests is_lowercase => sub {
                my $self = shift;
                is( $self->letter, lc( $self->letter ), "Letter is lowercase" );
            };
        };

        done_testing;

   DECLARE
    Note: no need to shift $self, it is done for you!

    sample.t

        use strict;
        use warnings;

        use Fennec::Declare;

        sub letter {
            my $self = shift;
            ( $self->{letter} ) = @_ if @_;
            return $self->{letter};
        }

        describe letters {
            case a { $self->letter('a') }

            case b { $self->letter('b') }

            case c { $self->letter('c') }

            case d { $self->letter('d') }

            tests is_letter {
                like( $self->letter, qr/^[a-z]$/i, "Got a letter" );
            }

            tests is_lowercase {
                is( $self->letter, lc( $self->letter ), "Letter is lowercase" );
            }
        }

        done_testing;

  MOCKING
    See Mock::Quick for more details

   OBJECT ON THE FLY
        my $obj = qobj(
            foo => 'foo',
            bar => qmeth { 'bar' },
            baz => sub { 'baz' },
        );

        is( $obj->foo, 'foo' );
        is( $obj->bar, 'bar' );
        is( ref $obj->baz, 'CODE', "baz is a method that returns a coderef" );

        # All methods autovivify as read/write accessors:
        lives_ok { $obj->blah( 'x' ) };

        # use qstrict() to make an object that does not autovivify accessors.

   SCOPE OF MOCKS IN FENNEC
    With vanilla Mock::Quick a mock is destroyed when the control object is
    destroyed.

        my $control = qtakeover Foo => (blah => 'blah');
        is( Foo->blah, 'blah', "got mock" );
        $control = undef;
        ok( !Foo->can('blah'), "Mock destroyed" );

        # WITHOUT FENNEC This issues a warning, the $control object is ignored so
        # the mock is destroyed before it can be used.
        qtakover Foo => (blah => 'blah');
        ok( !Foo->can('blah'), "Mock destroyed before it could be used" );

    With the workflow support provided by Fennec, you can omit the control
    object and let the mock be scoped implicitly.

        tests implicit_mock_scope => sub {
            my $self = shift;
            can_ok( $self, 'QINTERCEPT' );
            qtakeover Foo => (blah => sub { 'blah' });
            is( Foo->blah, 'blah', "Mock not auto-destroyed" );
        };

        describe detailed_implicit_mock_scope => sub {
            qtakeover Foo => ( outer => 'outer' );
            ok( !Foo->can( 'outer' ), "No Leak" );

            before_all ba => sub {

CHANGES
v2.019 - Jun 10 2018
    * Following NeilB's deprecation procedure

v2.018 - Jun 08 2018
    * Doc fixes, deprecate in favor of Test2::Suite

v2.017 - ?
    ?

v2.016 - Mar 18 2014
    * Bugfix for 2.015 issue

v2.015 - Mar 18 2014
    * Better subclass support

V2.014 - Mar 17 2014
    * Support FENNEC_PARALLEL=0 env var.

V2.009 - Better mocking scoping
 * Make it so that Mock::Quick control objects can be implicitly captured by
   Fennec scopes.

V2.008 - Supress messages on syntax check

V2.007 - Supress never run message on syntax check

V2.002-V2.006 - Minor Bugfixes

V2.001 - May 18 2013
    * Add 'class' import argument

v2.000 - May 18 2013
    * Replaced listener system with collector system
    * Removed evil auto-run hack
    * LOTS of bugfixes
    * Fixed win32 (!!!)
    * Added Child.pm as a default import
    * Lots of new features
    * Architecture overhaul
    * Removed dead code
    * Lots of simplifications
    * Additional testing

v1.014 - May 07 2013
    * Fixed more ordering issues

v1.013 - May 06 2013
    * Add 2 test files for badly broken behavior
    * Fixed broken behavior
     * Nested test structures did not run in proper separate procs
     * test_sort => 'ordered' did not run any tests

vXXXX - Lots
    * Missed a lot :-(

v1.008 - May 12, 2011 - Chad Granum:
    * Fix? hisenbug in TB listener.

v1.002-v1.007
    * Forgot to fill this out when I made the changes :-(

v1.001 - Feb 18, 2011 - Chad Granum:
    * Create shortcut for using runner, improve runner re-exec command
    * Remove verbose messages from STDERR, using STDOUT they cause less spam

v1.000 - Jan 24, 2011 - Chad Granum
    First stable release, API is now set.


js/fennec.js
$( function() {
    $( '.nav' ).click( function() {
        var item = $(this)

        $('#subnav ul').empty();
        $('ul.second_subnav').detach();
        $('#view').empty();
        $('#error_window').click( function() {
            $(this).hide();
        });

        jQuery.ajax(
            item.attr( 'id' ) + '.html',
            {
                dataType: 'html',
                success: function( data ) {
                    $( '.nav' ).removeClass( 'active' )
                    item.addClass( 'active' )

                    build_content( data );
                    fixView();
                },
                error: function() {
                    $( '#error_window' ).show();
                    $( '#error_window ul.errors' ).append( "<li>Error loading page</li>" )
                    $( '.nav' ).removeClass( 'active' )
                    item.addClass( 'active' )
                    fixView();
                }
            }
        )
    });

    var hash = window.location.hash
    if ( hash ) {
        var page = hash.split('-');
        $( page[0] ).trigger( 'click' )
    }
    else {
        $( '#fennec' ).trigger( 'click' )
    }

    var navstate = 1;
    $('#nonav').click( function() {
        if ( navstate ) {
            navstate = 0;
            $('#view').addClass( 'nonav' );
            $('#subnav').addClass( 'nonav' );
            $('ul.second_subnav').addClass( 'nonav' );
            $(this).addClass( 'nonav' );
            $(this).text( '' );
        }
        else {
            navstate = 1;
            $('#view').removeClass( 'nonav' );
            $('#subnav').removeClass( 'nonav' );
            $('ul.second_subnav').removeClass( 'nonav' );
            $(this).removeClass( 'nonav' );
            $(this).text( 'Hide Navigation' );
        }
    });
})

function fixView() {
    var height = $('#subnav').outerHeight();
    if ( height < 350 ) height = 350;
    $('#view').css( 'min-height', height );
}

function build_content( data ) {
    var view = $( '#view' );
    var subnav = $( 'ul#main_subnav' );

    var hash = window.location.hash;
    if ( !hash ) hash = '#fennec';
    var nav = hash.split('-');

    var new_stuff = $( '<div></div>' );
    new_stuff.html( data );
    $(new_stuff).find( 'dl.listnav' ).each( function() {
        $(this).children('dt').each( function() {
            var id = $(this).attr( 'id' );
            var dt = $(this);
            var dd = $(this).next();
            var navitem = $(
                '<li id="' + id + '"><a href="' + nav[0] + '-' + id + '">' + dt.html() + '</a></li>'
            );
            var classes = dd.attr( 'class' );
            var viewitem = $(
                '<div style="display: none" class="' + classes + '">' + dd.html() + '</div>'
            );

            process( id, viewitem );

            var normal_click = function() {
                view.children().hide();
                $('ul.second_subnav').hide();
                subnav.children().removeClass( 'active' );
                viewitem.show();
                var sn = $('ul#SN-' + id);
                if ( sn.length ) {
                    sn.show();
                    sn.children().removeClass('active');
                    if ( sn.children().length > 4 ) {
                        subnav.children().hide();
                        navitem.unbind( 'click' );
                        navitem.click( function() {
                            subnav.children().show();
                            navitem.unbind( 'click' );
                            navitem.click( normal_click );
                            fixView();
                        });
                    }
                    sn.children().first().click();
                }
                navitem.addClass( 'active' );
                navitem.show();
                fixView();
            };

            navitem.click( normal_click );

            subnav.append( navitem );
            view.append( viewitem );
        });

        if ( nav[1] ) {
            $( '#' + nav[1] ).trigger( 'click' );
            $('ul#SN-' + nav[1]).each( function() {
                $(this).show();
                if ( nav[2] ) {
                    $(this).find( '#' + nav[2] ).trigger( 'click' );
                }
            });
        }
        else {
            subnav.children().first().trigger( 'click' );
        }
    })
}

function process( id, container ) {
    container.find( 'div.symbol_list' ).each( function() {
        var list = $(this);
        jQuery.ajax(
            list.attr( 'src' ),
            {
                dataType: 'json',
                success: function( data ) {
                    list.replaceWith( build_symbol_list( data ));
                    fixView();
                },
                error: function(blah, message1, message2) {
                    $( '#error_window' ).show();
                    $( '#error_window ul.errors' ).append( "<li>Error loading " + list.attr( 'src' ) + "</li>" )
                    fixView();
                }
            }
        )
    });

    container.find( 'dl.sub_list' ).each( function() {
        $(this).detach();
        build_sub_list( id, $(this) );
        fixView();
    });

    process_samples( container );
}

function process_samples( container ) {
    container.find( 'script.code' ).each( function() {
        $(this).replaceWith( build_code( $(this).text() ));
    });

    container.find( 'div.code' ).each( function() {
        var list = $(this);
        jQuery.ajax(
            list.attr( 'src' ),
            {
                dataType: 'text',
                success: function( data ) {
                    list.replaceWith( build_code( data ));
                    fixView();
                    start_debugger();
                },
                error: function(blah, message1, message2) {
                    $( '#error_window' ).show();
                    $( '#error_window ul.errors' ).append( "<li>Error loading " + list.attr( 'src' ) + "</li>" )
                    fixView();
                }
            }
        )
    });

    container.find( 'div.output' ).each( function() {
        var list = $(this);
        jQuery.ajax(
            list.attr( 'src' ),
            {
                dataType: 'text',
                success: function( data ) {
                    list.replaceWith( build_output( data ));
                    fixView();
                    start_debugger();
                },
                error: function(blah, message1, message2) {
                    $( '#error_window' ).show();
                    $( '#error_window ul.errors' ).append( "<li>Error loading " + list.attr( 'src' ) + "</li>" )
                    fixView();
                }
            }
        )
    });

    container.find( 'div.vim' ).each( function() {
        var list = $(this);
        jQuery.ajax(
            list.attr( 'src' ),
            {
                dataType: 'text',
                success: function( data ) {
                    list.replaceWith( build_vim( data ));
                    fixView();
                    start_debugger();
                },
                error: function(blah, message1, message2) {
                    $( '#error_window' ).show();
                    $( '#error_window ul.errors' ).append( "<li>Error loading " + list.attr( 'src' ) + "</li>" )
                    fixView();
                }
            }
        )
    });
}

function build_code( data ) {
    var brush = new SyntaxHighlighter.brushes.Perl();

    brush.init({ toolbar: false });
    return brush.getHtml( data );
}

function build_output( data ) {
    var brush = new SyntaxHighlighter.brushes.TAP();

    brush.init({ toolbar: false });
    return brush.getHtml( data );
}

function build_vim( data ) {
    var brush = new SyntaxHighlighter.brushes.Vimscript();

    brush.init({ toolbar: false });
    return brush.getHtml( data );
}


function build_sub_list( pid, list ) {
    var subnav = $( '<ul id="SN-' + pid + '" style="display: none;" class="second_subnav listnav"></ul>' );

    var hash = window.location.hash;
    if ( !hash ) hash = '#fennec';
    var nav = hash.split('-');

    list.find( 'dt' ).each( function() {
        var navkey = $(this).text();
        var section = $(this).next();
        process_samples( section );
        build_sub_list_item( pid, navkey, section, nav, subnav );
    });

    $("#subnav").append( subnav );
}

function build_sub_list_item( pid, navkey, section, nav, subnav ) {
    var navitem = $(
        '<li id="' + navkey + '"><a href="' + nav[0] + '-' + pid + '-' + navkey + '">' + navkey + '</a></li>'
    );
    var viewitem = $(
        '<div style="display: none"><h2>' + navkey + '</h2></div>'
    );
    viewitem.append( section );

    navitem.click( function() {
        $('#view').children().hide();
        subnav.children().removeClass( 'active' );
        navitem.addClass( 'active' );
        viewitem.show();
        fixView();
    });

    subnav.append( navitem );
    $('#view').append( viewitem );
}

function build_symbol_list( data ) {
    var table = $( '<table class="symbol_list"><tbody><tr><th>Name</th><th>Description &nbsp;&nbsp; <small>(Click a row for usage details)</small</th></tr></tbody></table>' );
    for (key in data) {
        var name = data[key]['name'];
        if ( !name ) name = key;
        var row = $( '<tr class="symbol" onclick="expandDesc(this)"></tr>' );
        row.append( $('<td class="left">' + name + '</td>') );
        row.append( $('<td class="right">' + data[key]['desc'] + '</td>') );

        var details = $( '<td colspan="2"></td>' );
        if ( data[key]['usage'] ) {
            var list = $( '<ul class="usage"></ul>' );
            for ( i in data[key]['usage'] ) {
                var item = $( '<li>' + data[key]['usage'][i] + '</li>' );
                list.append( item );
            }
            details.append( list );
        }
        details.append( data[key]['details'] );

        var drow = $( '<tr class="symbol_details" style="display: none;"></tr>' );
        drow.append( details );

        table.append( row );
        table.append( drow );
    }

    return table;
}

function expandDesc( e ) {
    $(e).toggleClass( 'open' );
    $(e).next().toggle();
}

function openRole( role ) {
    $('#main_subnav').find('#roles').trigger( 'click' );
    $('#SN-roles').find('#' + role).trigger( 'click' );
    fixView();
}

function start_debugger() {
    stage_set_step(0);

    $('#stage_r').unbind( 'click' );
    $('#stage_b').unbind( 'click' );
    $('#stage_f').unbind( 'click' );
    $('#stage_e').unbind( 'click' );

    $('#stage_r').click( function() { stage_set_step(0) } );
    $('#stage_b').click( function() { stage_set_step(step - 1) } );
    $('#stage_f').click( function() { stage_set_step(step + 1) } );
    $('#stage_e').click( function() { stage_set_step(steps.length - 1) } );
}

steps = new Array(
    { line: 0,  out: 0,  proc: 'parent', i: "", s: "", b: "" },
    { line: 1,  out: 3,  proc: 'parent', i: "", s: "", b: "" },
    { line: 3,  out: 5,  proc: 'parent' },
    { line: 28, out: 6,  proc: 'parent',  i: '' },
    { line: 6,  out: 6,  proc: 'parent',  i: 0, s: '' },
    { line: 7,  out: 6,  proc: 'parent',  s: 0, b: '' },
    { line: 8,  out: 6,  proc: 'parent',  b: 100, i: 0 },
    { line: 10, out: 6,  proc: 'parent',  i: 1 },

    { line: 12, out: 6,  proc: 'child_1', s: 1, b: 100 },
    { line: 15, out: 6,  proc: 'child_1', b: 0 },
    { line: 18, out: 7,  proc: 'child_1' },
    { line: 19, out: 8,  proc: 'child_1' },
    { line: 20, out: 9,  proc: 'child_1', b: 0},
    { line: 21, out: 9,  proc: 'child_1', s: 1, i: 1, b: 5 },
    { line: 24, out: 10, proc: 'child_1', s: 0, i: 0, b: 5 },

    { line: 13, out: 10, proc: 'child_2', s: 2, b: 100 },
    { line: 15, out: 10, proc: 'child_2', b: 0 },
    { line: 18, out: 11, proc: 'child_2' },
    { line: 19, out: 12, proc: 'child_2' },
    { line: 20, out: 13, proc: 'child_2', b: 0 },
    { line: 21, out: 13, proc: 'child_2', s: 2, i: 1, b: 5 },
    { line: 24, out: 14, proc: 'child_2', s: 0, i: 0, b: 5 },

    { line: 25, out: 14, proc: 'parent' },
    { line: 29, out: 16, proc: 'parent' },
    { line: 0,  out: 17, proc: 'parent', i: 1, s: 0, b: 100 }
);

var step = 0;
var proc = 'parent';

function stage_set_step( num ) {
    // Sanity
    if ( num < 0 ) num = 0;
    if ( num >= steps.length ) num = steps.length - 1;

    step = num;
    var data = steps[step];
    console.log( data );

    // Hide children
    $( '#debugger tr' ).removeClass( 'active' );
    $( '#debugger tr.child' ).hide();
    // Show proc if not parent
    if ( data['proc'] != 'parent' ) {
        $( '#debugger tr.' + data['proc'] ).show();
    }
    if( data['line'] ) $( '#debugger tr.' + data['proc'] ).addClass( 'active' );

    // if proc changed, copy parent vals to proc
    if ( data['proc'] != proc ) {
        proc = data['proc'];
        if ( proc != 'parent' ) {
            var parent = $( '#debugger_state tr.parent'  );
            var child  = $( '#debugger_state tr.' + proc );
            child.find( 'td.vinit' ).text( parent.find( 'td.vinit' ).text()   );
            child.find( 'td.vstate' ).text( parent.find( 'td.vstate' ).text() );
            child.find( 'td.vbleed' ).text( parent.find( 'td.vbleed' ).text() );
        }
    }
    // Set vals
    var prow = $( '#debugger_state tr.' + proc );
    if ( 'i' in data ) prow.find( 'td.vinit'  ).text( data['i'] );
    if ( 's' in data ) prow.find( 'td.vstate' ).text( data['s'] );
    if ( 'b' in data ) prow.find( 'td.vbleed' ).text( data['b'] );

    // unhighlight lines
    $( 'td#debug_source div.syntaxhighlighter div.line' ).removeClass( 'debug_highlight' );

    // Highlight line
    $(
        'td#debug_source div.syntaxhighlighter div.line.number' + data['line']
    ).addClass( 'debug_highlight' );

    // Hide output
    $( 'td#debug_out div.syntaxhighlighter div.line' ).hide();

    // Display output up to 'out'
    for( var i = 1; i <= data['out']; i++ ) {
        $( 'td#debug_out div.syntaxhighlighter div.line.number' + i ).show();
    }
}

